fix(ai): throw error for missing start chunks in processUIMessageStream

codewarnab opened the pull request:
This PR addresses a runtime error: `TypeError: Cannot read properties of undefined (reading 'text')` in `processUIMessageStream`.

### Issue
The issue occurs when `text-delta`, `reasoning-delta`, or `tool-input-delta` chunks are received before their corresponding start chunks (e.g., `text-start`). This can happen in malformed streams, network issues causing packet reordering, or specific backend behaviors where a delta is emitted immediately without a start frame.

Previously, `processUIMessageStream` assumed that a start part (initialized in `activeTextParts`, etc.) always existed when a delta arrived. When it didn't, accessing properties like `textPart.text` caused a crash.

### Fix
This change adds explicit null checks for the active parts (`textPart`, `reasoningPart`, `partialToolCall`). If a delta is received for a missing part, it now throws a descriptive error:
> "Received text-delta for missing text part with ID "text-1". Ensure a "text-start" chunk is sent before any "text-delta" chunks."

This descriptive error aids in debugging the actual upstream issue (broken stream) rather than crashing the UI with an opaque TypeError.

## Changes
- **packages/ai/src/ui/process-ui-message-stream.ts**: Added guarding `if (part == null)` checks in `text-delta`, `reasoning-delta`, and `tool-input-delta` cases.
- **packages/ai/src/ui/process-ui-message-stream.test.ts**:
    - Added test cases simulating malformed streams (delta without start) for text, reasoning, and tool-inputs.
    - Verified that specific descriptive errors are thrown.
    - Updated test harness `consumeStream` to properly propagate errors for validation.
    
## Related Issues

Fixes #11700

codewarnab committed:
fix(ai): handle missing start chunks in processUIMessageStream

codewarnab committed:
Merge branch 'main' into fix/process-ui-stream-null-parts-clean

codewarnab committed:
Merge branch 'main' into fix/process-ui-stream-null-parts-clean

codewarnab committed:
Merge branch 'main' into fix/process-ui-stream-null-parts-clean

codewarnab committed:
Merge branch 'main' into fix/process-ui-stream-null-parts-clean

aayush-kapoor committed:
Merge branch 'main' into fix/process-ui-stream-null-parts-clean

codewarnab committed:
Merge branch 'main' into fix/process-ui-stream-null-parts-clean

codewarnab commented on packages/ai/src/ui/process-ui-message-stream.test.ts:
```diff
@@ -224,6 +224,105 @@ describe('processUIMessageStream', () => {
+      ).rejects.toThrow(
+        'Received tool-input-delta for missing tool call with ID "tool-1". ' +
+        'Ensure a "tool-input-start" chunk is sent before any "tool-input-delta" chunks.',
+      );
```
Not an issue - The onError callback explicitly re-throws (throw error), so the error propagates correctly to rejects.toThrow().

codewarnab committed:
Merge branch 'main' into fix/process-ui-stream-null-parts-clean

codewarnab committed:
Merge branch 'main' into fix/process-ui-stream-null-parts-clean

codewarnab committed:
fix(ui): add tests for missing text/reasoning start chunks

codewarnab commented:
> can you first address all the comments of the above bots?

addressed all of them

codewarnab committed:
feat(ai): add UIMessageStreamError for malformed stream chunks

codewarnab commented:
> does this break backwards compat, i.e. are there cases where it might have worked before but would error after this change?

**No breaking changes** - This change strictly improves error handling without breaking any legitimate use cases.
**Analysis:**
| Scenario | Before this PR | After this PR |
|----------|----------------|---------------|
| **Valid stream** (with proper `*-start` chunks) | ✅ Works | ✅ Works (unchanged) |
| **Malformed stream** (missing `*-start` chunks) | ❌ Crashes with unhelpful `TypeError: Cannot read properties of undefined (reading 'text')` | ✅ Throws descriptive `UIMessageStreamError` with chunk type/ID |

**Why it's not a breaking change:**
- The old behavior was a crash (bug), not intended functionality
- No consumer could have depended on receiving a `TypeError` specifically
- The new error extends `AISDKError`, maintaining SDK error conventions
- Error messages are preserved—only the error type has been upgraded
**The only "break" is:** Code that previously crashed silently will now surface clear, actionable errors. This is the intended improvement.

codewarnab committed:
Merge branch 'main' into fix/process-ui-stream-null-parts-clean

aayush-kapoor committed:
Merge branch 'main' into fix/process-ui-stream-null-parts-clean

codewarnab committed:
Merge branch 'main' into fix/process-ui-stream-null-parts-clean

aayush-kapoor commented:
hold off on merging this - there might be a different problem than the original for which the PR was raised

aayush-kapoor committed:
cs change

Previously, receiving certain malformed chunks would crash with an unhelpful TypeError. Now throws a dedicated UIMessageStreamError with clear, actionable error messages that include the chunk type and ID for easier debugging.

aayush-kapoor committed:
Merge branch 'main' into fix/process-ui-stream-null-parts-clean

aayush-kapoor reviewed (changes requested):
run `pnpm prettier-fix` to fix failing CI

aayush-kapoor committed:
pretty
